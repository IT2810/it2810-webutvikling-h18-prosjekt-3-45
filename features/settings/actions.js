/**
 * Set whether or not the feature for a target amount of
 * daily todos is finished.
 */
export const setGoalEnabled = enabled => ({
  type: 'SET_GOAL_ENABLED',
  enabled,
});

/**
 * Set the target daily amount of todos to finish.
 */
export const setTodoGoal = goal => ({
  type: 'SET_TODO_GOAL',
  goal,
});
