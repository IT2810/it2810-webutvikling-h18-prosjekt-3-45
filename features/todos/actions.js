import uuidv4 from 'uuid/v4';

export const addTodo = (text, description, date) => ({
  type: 'ADD_TODO',
  id: uuidv4(),
  text,
  description,
  date,
});

export const addPedometerTodo = (text, description, date, stepsGoal) => ({
  type: 'ADD_TODO',
  id: uuidv4(),
  creationDate: new Date().toGMTString(),
  isPedometer: true,
  text,
  description,
  date,
  stepsGoal,
});

export const updateTodo = (id, data) => ({
  type: 'UPDATE_TODO',
  id,
  data,
});

export const finishTodo = id => ({
  type: 'FINISH_TODO',
  id,
});

export const deleteTodo = id => ({
  type: 'DELETE_TODO',
  id,
});
