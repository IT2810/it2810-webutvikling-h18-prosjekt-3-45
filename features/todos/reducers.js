const todo = (state = {}, action) => {
  switch (action.type) {
    // Handle adding of todos and the undoing of deletion of todos.
    case 'ADD_TODO':
    case 'READD_TODO':
      // Simply add the specified information into an object, with some
      // additional default information.
      return {
        ...state,
        text: action.text,
        description: action.description,
        date: action.date,
        id: action.id,
        done: false,

        // These apply only when tracking the pedometer
        creationDate: action.creationDate,
        isPedometer: action.isPedometer,
        stepsGoal: action.stepsGoal,
      };

    case 'UPDATE_TODO':
      // Do not update the todo if the ID does not match.
      if (action.id !== state.id) {
        return state;
      }

      return {
        // Merge the old and the new data.
        ...state,
        ...action.data,
      };

    case 'FINISH_TODO':
      // Do not finish this todo if the ID does not match.
      if (action.id !== state.id) {
        return state;
      }

      return {
        // Mark the todo as finished.
        ...state,
        done: true,
        finished: action.finished,
      };

    default:
      return state;
  }
};

const todos = (state = [], action) => {
  switch (action.type) {
    case 'ADD_TODO':
      // Add the new todo to the bottom of the list of todos.
      return [...state, todo(undefined, action)];

    case 'READD_TODO':
      // Remove the todo which is marked as finished from the list.
      const filtered = state.filter(todo => todo.id !== action.id);

      return [
        // Readd the old todo in the index in which it used to be.
        ...filtered.slice(0, action.index + 1),
        todo(undefined, action),
        ...filtered.slice(action.index + 1),
      ];

    case 'UPDATE_TODO':
    case 'FINISH_TODO':
      // For these actions, we simply map all todos and
      // pass them the action. The todos check their own
      // identificators and update accordingly.
      return state.map(entry => todo(entry, action));

    case 'DELETE_TODO':
      // Filter away all todos whichc have a different ID than the one to delete.
      return state.filter(entry => entry.id !== action.id);

    case 'REHYDRATE':
      // When data is loaded from AsyncStorage, we add these to the start of
      // the list. Usually, this should mean the new list is equal to the loaded list,
      // but we merge it with the existing todos in case AsyncStorage is really slow,
      // so that the user is somehow able to create extra todos before loading finishes.
      return [...action.state.todos, ...state];

    default:
      return state;
  }
};

export default todos;
