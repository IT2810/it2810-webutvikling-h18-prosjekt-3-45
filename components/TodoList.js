import React, { Component, Fragment } from 'react';
import { connect } from 'react-redux';
import { View, Alert } from 'react-native';
import Todo from './Todo';
import PedometerTodo from './PedometerTodo';
import { Toast } from 'native-base';
import store from '../store';
import SchemaModal from './TodoAdder/SchemaModal';
import { finishTodo, deleteTodo, updateTodo } from '../features/todos/actions';

// This class is exported both as the default export and as a component
// wrapped using connect to ease testing of this component.
export class TodoList extends Component {
  /**
   * modalVisible is whether the modal for adding a new todo should be visible
   * pressedTodo is the todoobject (containing the states) that is pressed
   * @type {{modalVisible: boolean, pressedTodo: {}}}
   */
  state = {
    modalVisible: false,
    pressedTodo: {},
  };

  /**
   * Handles a press on the UNDO button in the toast which is displayed
   * after swiping away a TODO.
   */
  handleUndo = (reason, todo, index) => {
    if (reason === 'user') {
      // Strip out the done flag and finished time, as we want
      // this to be reset by the reducer.
      const { done, finished, ...rest } = todo;

      store.dispatch({
        // This action type is a special action which allows us to
        // insert a TODO at a specified index.
        type: 'READD_TODO',
        ...rest,
        index,
      });
    }
  };

  /**
   * Handles sliding of todos
   * @param direction   the direction of the sliding
   * @param id          id of the todo
   */
  handleOpen = (direction, id, close) => {
    if (direction === 'left') {
      // Find the TODO and save it so that it can be restored later.
      const index = this.props.todos.findIndex(todo => todo.id === id);

      const todo = this.props.todos[index];

      handleFinishTodo = (todo, index) => {
        Toast.show({
          text: 'Todo marked as finished',
          buttonText: 'Undo',
          onClose: reason => this.handleUndo(reason, todo, index),
          duration: 5000,
        });

        this.props.finishTodo(id);
      };

      if (!todo.done && todo.isPedometer) {
        Alert.alert(
          'Are you sure?',
          "You don't have the required amount of steps!",
          [
            { text: 'OK', onPress: () => handleFinishTodo(todo, index) },
            { text: 'Cancel', onPress: () => close(), style: 'cancel' },
          ],
          { cancelable: true },
        );
        return;
      }

      handleFinishTodo(todo, index);

      close();
    }
  };

  /**
   * When a todo is pressed in the list, find the states and send these to updateTodo
   * @param id
   */
  handlePress = id => {
    // Finds the id of pressed todo among todos
    this.state.pressedTodo = this.props.todos.find(todo => todo.id === id);
    // Shows the form
    this.setModalVisibility(true);
  };

  /**
   * Deletes the todo from the list
   * @param id      id for redux
   */
  handleDelete = id => {
    this.props.deleteTodo(id);
  };

  /**
   * Sets the visibility of the form modal to the visible-state
   * @param visible   boolean (whether or not the modal should be visible)
   */
  setModalVisibility = visible => {
    this.setState({
      modalVisible: visible,
    });
  };

  /**
   * Updates a new todo with forminput from editing todo
   * This will display the todoschema s.t. we can get a form modal
   * @param data        object containing text, description and date for the changed todo
   */
  updateTodo = data => {
    this.setModalVisibility(false);
    this.props.updateTodo(this.state.pressedTodo.id, data);
  };

  render() {
    return (
      <Fragment>
        <View>
          <SchemaModal
            saveForm={this.updateTodo}
            setModalVisibility={this.setModalVisibility}
            modalVisible={this.state.modalVisible}
            currentTodo={this.state.pressedTodo}
          />
        </View>
        {this.props.todos.map(
          todo =>
            todo.isPedometer ? (
              <PedometerTodo
                key={todo.id}
                onOpen={this.handleOpen}
                onDelete={this.handleDelete}
                onPress={this.handlePress}
                todo={todo}
              />
            ) : (
              <Todo
                key={todo.id}
                onOpen={this.handleOpen}
                onDelete={this.handleDelete}
                onPress={this.handlePress}
                todo={todo}
              />
            ),
        )}
      </Fragment>
    );
  }
}

const mapDispatchToProps = {
  finishTodo,
  deleteTodo,
  updateTodo,
};

export default connect(
  undefined,
  mapDispatchToProps,
)(TodoList);
