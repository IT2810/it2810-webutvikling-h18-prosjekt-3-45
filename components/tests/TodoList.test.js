import React from 'react';
import renderer from 'react-test-renderer';
import { TodoList } from '../TodoList';
import todos from '../../features/todos/reducers';
import { addTodo } from '../../features/todos/actions';

const mockTodos = todos(undefined, addTodo('Test todo'));

// snapshot test for completed todolist
describe('TodoList', () => {
  const tree = renderer
    .create(<TodoList showDone={true} todos={mockTodos} />)
    .toJSON();

  it('should render correctly', () => {
    expect(tree).toMatchSnapshot();
  });
});

// snapshot test for incompleted todolist
describe('TodoList', () => {
  const tree = renderer
    .create(<TodoList showDone={false} todos={mockTodos} />)
    .toJSON();

  it('should render correctly', () => {
    expect(tree).toMatchSnapshot();
  });
});
