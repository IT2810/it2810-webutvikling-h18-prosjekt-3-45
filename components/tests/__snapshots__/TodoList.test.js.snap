// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoList should render correctly 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#fff",
      "flex": 1,
      "height": 1334,
    }
  }
>
  <View>
    <Modal
      animationType="slide"
      hardwareAccelerated={false}
      onRequestClose={[Function]}
      transparent={false}
      visible={false}
    >
      <View
        style={
          Object {
            "flex": 1,
          }
        }
      >
        <View
          style={
            Object {
              "backgroundColor": "#fff",
              "flex": 1,
              "height": 1334,
            }
          }
        >
          <View
            style={
              Array [
                Object {
                  "backgroundColor": "#F8F8F8",
                  "borderColor": "#cbcbcb",
                  "borderTopWidth": 0.5,
                  "elevation": 3,
                  "flexDirection": "row",
                  "height": 55,
                  "justifyContent": "center",
                  "left": 0,
                  "paddingBottom": 0,
                  "right": 0,
                },
                Object {
                  "backgroundColor": "#EEE",
                  "borderColor": "#CCC",
                  "borderTopWidth": 0.5,
                  "flexDirection": "row",
                  "justifyContent": "space-between",
                  "paddingHorizontal": 20,
                  "paddingVertical": 5,
                },
              ]
            }
            transparent={true}
          >
            <View
              accessible={true}
              isTVSelectable={true}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                Object {
                  "alignItems": "center",
                  "alignSelf": "flex-start",
                  "backgroundColor": "#b5b5b5",
                  "borderBottomWidth": null,
                  "borderColor": "#007aff",
                  "borderLeftWidth": null,
                  "borderRadius": 5,
                  "borderRightWidth": null,
                  "borderTopWidth": null,
                  "borderWidth": undefined,
                  "elevation": 2,
                  "flexDirection": "row",
                  "height": 45,
                  "justifyContent": "space-between",
                  "opacity": 1,
                  "paddingBottom": 6,
                  "paddingTop": 6,
                  "shadowColor": undefined,
                  "shadowOffset": undefined,
                  "shadowOpacity": undefined,
                  "shadowRadius": undefined,
                }
              }
            >
              <Text
                accessible={true}
                allowFontScaling={true}
                ellipsizeMode="tail"
                style={
                  Object {
                    "backgroundColor": "transparent",
                    "color": "#fff",
                    "fontFamily": "System",
                    "fontSize": 16.5,
                    "marginLeft": 0,
                    "marginRight": 0,
                    "paddingLeft": 16,
                    "paddingRight": 16,
                  }
                }
                uppercase={false}
              >
                Save
              </Text>
            </View>
            <View
              accessible={true}
              isTVSelectable={true}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                Object {
                  "alignItems": "center",
                  "alignSelf": "flex-start",
                  "backgroundColor": "#d9534f",
                  "borderBottomWidth": null,
                  "borderColor": "#007aff",
                  "borderLeftWidth": null,
                  "borderRadius": 5,
                  "borderRightWidth": null,
                  "borderTopWidth": null,
                  "borderWidth": undefined,
                  "elevation": 2,
                  "flexDirection": "row",
                  "height": 45,
                  "justifyContent": "space-between",
                  "opacity": 1,
                  "paddingBottom": 6,
                  "paddingTop": 6,
                  "shadowColor": undefined,
                  "shadowOffset": undefined,
                  "shadowOpacity": undefined,
                  "shadowRadius": undefined,
                }
              }
            >
              <Text
                accessible={true}
                allowFontScaling={true}
                ellipsizeMode="tail"
                style={
                  Object {
                    "backgroundColor": "transparent",
                    "color": "#fff",
                    "fontFamily": "System",
                    "fontSize": 16.5,
                    "marginLeft": 0,
                    "marginRight": 0,
                    "paddingLeft": 16,
                    "paddingRight": 16,
                  }
                }
                uppercase={false}
              >
                Cancel
              </Text>
            </View>
          </View>
        </View>
      </View>
    </Modal>
  </View>
</View>
`;
